app/controllers/application_controller/filter.rb

```diff
@@ -0,0 +1,152 @@
+class ApplicationController
+  class Filter
+    INTEGER_COMPARISON_KEYWORDS = ["eq", "gt", "gte", "lt", "lte", "nil", "not_nil"].freeze
+    STRING_COMPARISON_KEYWORDS  = ["contains", "eq", "starts_with", "ends_with", "nil", "not_nil"].freeze
+
+    attr_reader :apply, :api_doc_definition
+
+    def initialize(model, raw_filter, api_doc_definition)
+      self.query          = model
+      @raw_filter         = raw_filter
+      @api_doc_definition = api_doc_definition
+    end
+
+    def apply
+      return query if @raw_filter.blank?
+      @raw_filter.each do |k, v|
+        with_attribute_for_key(k) do |attribute|
+          if attribute["type"] == "string"
+            type = determine_string_attribute_type(attribute)
+            send(type, k, v)
+          else
+            errors.add("unsupported attribute type for: #{k}")
+          end
+        end
+      end
+
+      raise ApplicationController::Filter::Error.new(:error_document => errors) unless errors.blank?
+      query
+    end
+
+    private
+
+    attr_accessor :query
+
+    class Error < ArgumentError
+      attr_reader :error_document
+
+      def initialize(attrs)
+        @error_document = attrs[:error_document]
+      end
+    end
+
+    def with_attribute_for_key(key)
+      attribute = api_doc_definition.properties[key.to_s]
+      attribute ? yield(attribute) : errors.add("found unpermitted parameter: #{key}")
```

Comment:
If you have an error here, then it just skips the yield?  If so, won't that allow add_filter to continue on?

Reply:
I'm curious why this is in yield-style?  Seems just as easy for this method to return the value and let the caller do whatever it needs, since there's no "cleanup" afterwards that a block would be needed for.

Reply:
Yeah, this is one of the things I didn't get to cleaning up.

Reply:
Fixed.